Ambient wrapper chart

bleggett opened the pull request:
**Please provide a description of this PR:**

This is a purely cosmetic wrapper chart that is designed to make the docs here: https://istio.io/latest/docs/ambient/install/helm-installation/ and here https://istio.io/latest/docs/ambient/upgrade/helm-upgrade/ less intimidating.

1. If you want `ambient`, you always need this baseline. Making people always install each baseline component piecemeal just to use ambient is off-putting/unnecessarily intimidating.
2. This does nothing but compose the granular charts we already have and publish, so you are still free to install each component individually (and in fact we can still advise this for production deployments, Argo, whatever) - but we don't want people avoiding Helm installs because our quickstart has a bunch of unnecessary steps.
3. This is pretty trivial to add, and no harder to keep updated than the current `helm-profiles` - the biggest change is the required `release-builder`: https://github.com/istio/release-builder/pull/1849 update.

bleggett committed:
Just add an ambient umbrella chart

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett commented on manifests/charts/ambient/Chart.yaml:
```diff
@@ -0,0 +1,29 @@
+type: application
+version: 1.0.0
+appVersion: 1.0.0
+kubeVersion: ">= 1.23.0-0"
```
```suggestion
kubeVersion: ">= 1.23.0-0"
```

This can be removed if it's controversial (and maybe automatically updated somehow on release), but it mimics our documented k8s requirements and is a nice safeguard.

howardjohn replied:
Does it work with `helm template`? I don't know what "work" means exactly, I suppose that it ignores it entirely?

IIRC `helm template` will bake in some arbitrary version, wondering if they actually are using that here which would be a bit silly. Also probably fine since its presumably above 1.23, but a bit odd to effectively be accidentally doing a helm version check instead of a kube version check for that case

bleggett replied:
`helm template` has a `--kube-version` flag for this, but IDK what the default is

istio-testing commented:
@bleggett: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 6366ec358b986d5c44d65e5161e822b73ac7355f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51174/release-notes_istio/1792953800542130176) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett committed:
Relnote

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Fixup

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
We don't have this for other charts, so drop it

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

howardjohn commented on manifests/charts/ambient/values.yaml:
```diff
@@ -0,0 +1,32 @@
+# Copied from `helm-profiles/ambient.yaml`
```
can you not just do literally `profile: ambient`?

bleggett replied:
You cannot, but lemme make sure I'm not missing some non-obvious syntax.

bleggett replied:
s/cannot/can/g - TY!

howardjohn commented on manifests/charts/ambient/values.yaml:
```diff
@@ -0,0 +1,32 @@
+# Copied from `helm-profiles/ambient.yaml`
+
+global:
```
Helm noob question: does this end up as `global.variant` in the chart or does it flatten it to `variant`?

Or is it that each chart gets
* global -> global
* <chart name> -> flattened
?

bleggett replied:
`global.xxx` is never scoped/rescoped - it is always `global.xx` in all charts + all subcharts.

bleggett replied:
And if you redefine the global in a subchart:

> If a subchart declares a global variable, that global will be passed downward (to the subchart's subcharts), but not upward to the parent chart. There is no way for a subchart to influence the values of the parent chart.

> Also, global variables of parent charts take precedence over the global variables from subcharts.

howardjohn commented on manifests/charts/ambient/values.yaml:
```diff
@@ -0,0 +1,32 @@
+  variant: distroless
+
+# Overrides for the `istiod-remote` dep
+istiod:
```
I think helm can do some enablement things. Do we want that here?

base, cni, and ztunnel are ~singletons so having them together is fairly reasonable. Prod usage of istiod should use revisions though, so it seems like you would want to actually have `ambient` and `istiod-<rev>` for most prod usage?

Unless we are claiming this is just for simple use cases and prod should always fully separate?

bleggett replied:
You could do that with `helm install ambient --set istiod.revision`

howardjohn replied:
doesn't that remove the old one instantly though

bleggett replied:
it shouldn't, but lemme check.

howardjohn replied:
If not, what is managed the old revision?

howardjohn replied:
don't we need a `enabled` condition in chart.yaml then to allow to later remove the Istiod original

bleggett replied:
Yeah good call - I can add a dependency conditional.

howardjohn replied:
tbh might be good to just pre-emptively do it for all of them. But we can also do it on-demand if someone asks

bleggett replied:
Yeah latter preferred, since none of the others support revisions. 

I would kinda want to have eyeballs on decisions for making the other two conditional for $reason, so prefer to defer-as-requested.

howardjohn reviewed (commented):
Overall looks good, just a few questions

bleggett committed:
derp

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Conditionalize `istiod` to make managing revisions easier

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

howardjohn commented on manifests/charts/ambient/Chart.yaml:
```diff
@@ -0,0 +1,30 @@
+  - name: istiod
+    version: 1.0.0
+    repository: "file://../istio-control/istio-discovery"
+    condition: enable.istiod
```
```suggestion
    condition: istiod.enabled
```
?

I think I have seen this pattern before. If I am imagining it then this works too

howardjohn reviewed (approved):
LGTM.

Does this get published to the helm repos as well? automatically or with a release-builder tweak?

dhawton commented:
> LGTM.
> 
> Does this get published to the helm repos as well? automatically or with a release-builder tweak?

It does need a release-builder tweak if this is its own chart: https://github.com/istio/release-builder/blob/master/pkg/build/helm.go#L49-L70

bleggett commented:
Will do doc updates too.

istio-testing commented:
In response to a cherrypick label: new pull request created: #51771

bleggett commented:
Got a request to backport this to 1.22.x so it gets published: https://istio.slack.com/archives/C041EQL1XMY/p1719508844013239?thread_ts=1719420472.445439&cid=C041EQL1XMY

Docs will lag, but that's fine.

costinm reviewed (commented):
So why not a 'non-ambient' umbrella chart - and we are back to 5 years ago ? 
Looks like this chart also installs regular istio - should we add a bunch of other combos ? 

It is super easy to add more install methods  - we have a long history, removing them and dealing with user confusion is the hard part.

costinm commented:
Actually - the main reason I commented: since we moved the operator to 'ecosystem', can we do the same with this chart - and restore the "2 install methods for Istio" in the main tree ?  Perfectly fine to have 20 installers and charts and combos in ecosystem, each with their doc - I don't want 6 months from now to go back to 'people installed with the easy way in prod and now they can't upgrade safely. 

Or alternatively - deprecate istioctl install, since this chart also installs all istio components in the 'easy' but not production operable way. 

My main concern really is that we have too many install methods in the main tree ( plus dozens of vendor specific ones probably)

bleggett commented:
> Actually - the main reason I commented: since we moved the operator to 'ecosystem', can we do the same with this chart - and restore the "2 install methods for Istio" in the main tree ? Perfectly fine to have 20 installers and charts and combos in ecosystem, each with their doc.

Since this is basically 5 lines of YAML (a triviality), and I am personally more interested in having a way to test/validate our proper individual charts _can_ be composed in a normative Helm fashion without breaking, more than I am advocating everyone use a wrapper chart, I think it's worth bringing that up in WG - I honestly don't have a strong opinion either way there. I mostly want this to simplify quickstart/tirekick docs.
 
> Or alternatively - deprecate istioctl install, since this chart also installs all istio components in the 'easy' but not production operable way.

It's a bit unclear how we want `istioctl install` to work with Helm other than we are all agreed it should be as dead-simple a wrapper around `helm install` as is humanly possible - so probably best to take things one step at a time - at least, dropping the operator and properly tackling `istioctl install` as-a-helm-wrapper are the immediate steps.


> My main concern really is that we have too many install methods in the main tree ( plus dozens of vendor specific ones probably)

I share this concern - the only reason I added the wrapper chart is because it _literally_ does almost nothing - no values, no opinions, just a compositional effort to prove our charts can be safely composed - the instant we start adding anything to this wrapper chart (local values overrides, templates, etc) I would be in favor of booting it out, yeah.


> I don't want 6 months from now to go back to 'people installed with the easy way in prod and now they can't upgrade safely.

On this point (since I agree with it) can you bring your thoughts to bear on https://github.com/istio/istio.io/pull/15279 ? A lot of this boils down to:

- what we document
- how we message it to people

So I would very much like your opinions on that end of things, as it will clarify this end of things.


> Or alternatively - deprecate istioctl install, since this chart also installs all istio components in the 'easy' but not production operable way.
> 
> My main concern really is that we have too many install methods in the main tree ( plus dozens of vendor specific ones probably)
